src/device/pf_block_writer.c

```diff
@@ -4109,3 +4128,441 @@ void pf_put_pdport_data_adj (
+   pf_put_uint32 (
+      is_big_endian,
+      RPC_EPM_ANNOTATION_OFFSET,
+      res_len,
+      p_bytes,
+      p_pos);
+
+   pf_put_uint32 (
+      is_big_endian,
+      RPC_EPM_ANNOTATION_LENGTH,
+      res_len,
+      p_bytes,
+      p_pos);
+
+   pf_put_rpc_epm_annotation (p_tower->p_im_0_data, res_len, p_bytes, p_pos);
+
+   pf_put_uint32 (is_big_endian, RPC_EPM_FLOOR_LENGTH, res_len, p_bytes, p_pos);
+
+   pf_put_uint32 (is_big_endian, RPC_EPM_FLOOR_LENGTH, res_len, p_bytes, p_pos);
+
+   pf_put_rpc_floor_count (
+      is_big_endian,
+      RPC_TOWER_FLOOR_COUNT,
+      res_len,
+      p_bytes,
+      p_pos);
+
+   pf_put_rpc_floor_1_uuid (
+      is_big_endian,
+      &p_tower->floor_1,
+      res_len,
+      p_bytes,
+      p_pos);
+
+   pf_put_rpc_floor_2_uuid (
+      is_big_endian,
+      &p_tower->floor_2,
+      res_len,
+      p_bytes,
+      p_pos);
+
+   pf_put_rpc_floor_3_rpc (
+      is_big_endian,
+      &p_tower->floor_3,
+      res_len,
+      p_bytes,
+      p_pos);
+
+   pf_put_rpc_floor_4_udp (
+      is_big_endian,
+      &p_tower->floor_4,
+      res_len,
+      p_bytes,
+      p_pos);
+
+   pf_put_rpc_floor_5_ip (
+      is_big_endian,
+      &p_tower->floor_5,
+      res_len,
+      p_bytes,
+      p_pos);
+}
+
+/**
+ * Put rpc handle into a buffer
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_handle         In:   Rpc handle
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_handle (
+   bool is_big_endian,
+   const pf_rpc_handle_t * p_handle,
+   uint16_t res_len,
+   uint8_t * p_bytes,
+   uint16_t * p_pos)
+{
+   pf_put_uint32 (
+      is_big_endian,
+      p_handle->rpc_entry_handle,
+      res_len,
+      p_bytes,
+      p_pos);
+
+   pf_put_rpc_uuid (
+      is_big_endian,
+      &p_handle->handle_uuid,
+      res_len,
+      p_bytes,
+      p_pos);
+}
+
+/**
+ * Put rpc epm entry into a buffer. If no no protocol tower is
+ * configured / actual_count == 0, a empty entry is written.
+ * @param is_big_endian    In:   Endianness of the destination buffer.
+ * @param p_entry          In:   Rpc epm entry including protocol
+ *                               tower configuration.
+ * @param res_len          In:   Size of destination buffer.
+ * @param p_bytes          Out:  Destination buffer.
+ * @param p_pos            InOut:Position in destination buffer.
+ */
+static void pf_put_rpc_epm_entry (
+   bool is_big_endian,
+   const pf_rpc_entry_t * p_entry,
+   uint16_t res_len,
+   uint8_t * p_bytes,
+   uint16_t * p_pos)
+{
+   pf_put_uint32 (is_big_endian, p_entry->max_count, res_len, p_bytes, p_pos);
+
+   pf_put_uint32 (is_big_endian, p_entry->offset, res_len, p_bytes, p_pos);
+
+   pf_put_uint32 (is_big_endian, p_entry->actual_count, res_len, p_bytes, p_pos);
+
+   /*Add entry if there if not empty */
```

Comment:
`/* Add entry if not empty */`

Reply:
done

Reply:
also added docstrings for pf_put_rpc_uuid but can't figure out how to respond to it